### ISSUE TITLE
wrapper Autogenerated cmake list clash with openMM/noOpenMM support

### ISSUE THREAD

**Issue by ppxasjsm** [michellab/Sire#54]
Currently the free_en_analysis branch supports the compilation of sire with both openMM and without openMM. However, if any of the OpenMM wrappers are changed, by using the automated wrapper generation script, wrapper/Move/CMakeAutogenFile.txt is updated and will be included in the CMake list and contain OpenMM files in the PYPP_SOURCES variable. For that purpose I created two additional files CMakeNoOpenMM.txt and CMakeOpenMM.txt. These will populate the SIRE_OPENMM_WRAPPERS variable and are selected based on an if statement that check if OpenMM support is turned on or not. The script that updated  wrapper/Move/CMakeAutogenFile.txt, needs to respect that if the OpenMM/noOpenMM support is meant to be given. 
In general I am not sure how checking for OpenMM support is done currently is the best way of doing it. Certainly there will be no warning, if somd is attempted to be called for example. 
This again I think requires some tidying of the Cmake files as well as warning within the repository. I definitely need some help/discuss how to best approach this since I have very limited experience with this kind of thing.

**Comment by chryswoods** [comment]
I think that we need to differentiate between different "use" cases:

(1) A developer regenerating the wrappers.
(2) A developer compiling Sire and the wrappers from source
(3) A user running somd from a downloaded binary.

== For case (1) ==
Ideally, the same wrappers should be generated whether or not OpenMM is present. We want to generate the same set of wrappers, which inside themselves contains a switch to trigger whether or not to compile against OpenMM. So, you are correct that we need to update the script that generates the wrappers to know that it should not place the OpenMM-dependent wrappers into CmakeAutogenFile.txt, but that they should instead go into CMakeOpenMM.txt or CMakeNoOpenMM.txt. This can be fixed by editing wrapper/AutoGenerate/create_wrappers.py, near the bottom where it creates the CMakeAutogenFile.txt - line that reads FILE = open("CMakeAutogenFile.txt", "w"). At the moment it just globs up all of the .pypp.cpp files and adds them to CMakeAutogemFile.txt. You will need to add in a test to say that it a filename is contained in a list, then it should be written to CMakeOpenMM.txt, e.g.

openmm_files = [ "file1_openmm.pypp.cpp", "file2_openmm.pypp.cpp" etc. ]

for pyppfile in pyppfiles:
    if pyppfile in openmm_files:
       print >>OPENMM_FILE,"    %s" % pyppfile
    else:
       print >>FILE, "    %s" % pyppfile

== For case (2) ==

We need to have a better way for detecting whether or not to compile OpenMM at runtime, and also make sure that the user can override this selection (e.g. if they have OpenMM installed, but don't want to compile). This code is messy, if I remember correctly, and needs work.

== For case (3) ==

At the moment, somd is available regardless of whether or not OpenMM is supported. We need to decide whether we either:

(a) delete somd from the binary if OpenMM is not available, thereby making it clear that you need OpenMM to use somd, or 
(b) leave somd, but make sure that there is a good error message if OpenMM is not available, e.g.

./somd .....

=== Error. Sorry, but somd requirest OpenMM to be supported. To use OpenMM you must download an OpenMM-capable version of Sire and have OpenMM support on your machine. For more information, please see http://siremol.org/help/openmm_support ===

**Comment by ppxasjsm** [comment]
I agree with what you said. 

I don't mind how any of this is implemented at the moment. I think a -DSIRE_OPENMM=on, or something like that would be good. Something along those lines kind of exists, but it is not all that clear to me. I am not sure what the best solution most cmake appropriate style for this is. 

Also, I must admit, that I haven't been using the wrapper script lately, because it was easier to edit a couple of things manually rather than starting the virtual machine for a 2 line edit. 

==Use case (3)==
I am against not including the binary. I would instead give an error message. There is still a substantial amount of work to be done on somd or rather the OpenMMMD.py module. I will hopefully work on this next week!

**Comment by ppxasjsm** [comment]
Oh also, you might have some issues with installing pip and setup tools, this is also fixed in the free_en_analysis branch. I'll hopefully merge that into devel later today.

**Comment by chryswoods** [comment]
Ok - I think that the pip and setup tools problems will be fixed by the use of miniconda.

Surprisingly we don't have a SIRE_OPENMM flag (at least as far as I can see). If OpenMM is found, then it is immediately switched on. I will create this flag in the CMakeLists.txt file as it is a pain to use OpenMM on my laptop (have a load of clashes from when I worked with Gaetano).

**Comment by ppxasjsm** [comment]
Yes this should also  be fixed using conda. 

OpenMM is in conda. You could just use that OpenMM version. Installation takes not even 2 minutes. 
```
conda config --add channels http://conda.binstar.org/omnia
conda install --yes openmm
```
And I guess you have to set the OPENMM_ROOT_DIR since that is the main variable the find OpenMM bit in the camke is looking for.

**Comment by chryswoods** [comment]
Ok - I will look into this with conda. I have just committed the code that added the SIRE_USE_OPENMM option in the top-level CMakeLists.txt file, which then feeds through all other files to make sure that a user can disable use of OpenMM if they don't want it. I have made a change to wrapper/Move/CMakeLists.txt too, to split CMakeAutogen to create a CMakeOpenMMFile.txt so that the OpenMM-dependent wrappers are not compiled if OpenMM is not available. I will work on a way to automate the creation of this file with the wrapper generation script...
